### ISSUE TITLE
BSTR marshalling

### ISSUE THREAD

**Issue by MichalStrehovsky** [dotnet/runtimelab#167]
Interop/PInvoke/Array/MarshalArrayAsField/AsByValArray/AsByValArrayTest

**Comment by kant2002** [comment]
I have a question about current implementation.
In the shared marshallers code I see that when using READYTORUN codepath used marshalling from `System.StubHelpers.AnsiBSTRMarshaler` class when ILC code-path used marshalling in the `InteropHelpers` class.
Why difference exists?

**Comment by kant2002** [comment]
Other question which I forgot to ask. Is the BSTRMarshaller and other similar classes should belong to `Marshaller.cs` or in the `Marshaller.Aot.cs`?

**Comment by jkotas** [comment]
> Why difference exists?

We have not done refactoring required to unify them. I expect we will do such refactoring as part https://github.com/dotnet/runtimelab/tree/feature/DllImportGenerator project eventually: We will just make the helpers to follow the same shape as what we will be used for source generators.

**Comment by jkotas** [comment]
>  Is the BSTRMarshaller and other similar classes should belong to Marshaller.cs or in the Marshaller.Aot.cs?

The split between the two is not very clear. Just pick the one you like better.

**Comment by kant2002** [comment]
> We will just make the helpers to follow the same shape as what we will be used for source generators.

So I can try to implement just BSTR for ILC code paths for now as this is what would be path forward?

**Comment by jkotas** [comment]
Yes, that's fine

**Comment by MichalStrehovsky** [comment]
Marshaller.cs is also used in crossgen2. Crossgen2 only implements a subset of the marshallers to avoid bugs. We don't have a choice in NativeAOT since there's no JIT to fall back to. This one should probably be in .Aot.cs. I don't think crossgen2 is going to use it any time soon.

**Comment by kant2002** [comment]
What is `AnsiBSTR`? Is this same BSTR which allocated/deallocated using `SysAllocString / SysFreeString` just actual data not `wchar_t` but `char_t`?

**Comment by jkotas** [comment]
Yes. It have not seen it be ever used, except in a very old Visual Basic code.
